tests/unit/app_test.cc

```diff
@@ -0,0 +1,54 @@
+/*
+ * This file is open source software, licensed to you under the terms
+ * of the Apache License, Version 2.0 (the "License").  See the NOTICE file
+ * distributed with this work for additional information regarding copyright
+ * ownership.  You may not use this file except in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+/*
+ * Copyright (C) 2023 ScyllaDB
+ */
+
+#include <boost/test/tools/old/interface.hpp>
+#define BOOST_TEST_MODULE app_template
+
+#include <string>
+#include <boost/test/unit_test.hpp>
+#include <seastar/core/app-template.hh>
+
+using namespace seastar;
+
+// NOTE: only a single test case is hosted in this test, because the underlying
+// Seastar runtime does not do cleanup every bits when it tears down. and this
+// is a design decision at this moment, so even launching two seastar
+// applications in the same process sequentially is not supported.
+BOOST_AUTO_TEST_CASE(app_standard_memory_allocator) {
+    // by default, use conservative settings instead of maxing out the performance
+    // for testing app_template and underlying reactor's handling of different
+    // settings
+    app_template::seastar_options opts;
+    opts.smp_opts.thread_affinity.set_value(false);
+    opts.smp_opts.mbind.set_value(false);
+    opts.smp_opts.smp.set_value(1);
+    opts.smp_opts.lock_memory.set_value(false);
+    opts.smp_opts.memory_allocator = memory_allocator::standard;
+    opts.log_opts.default_log_level.set_value(log_level::error);
+    app_template app{std::move(opts)};
+    // app.run() takes `char**` not `char* const *`, so appease it
+    std::string prog_name{"prog"};
+    char* args[] = {prog_name.data()};
+    int exit_code = app.run(std::size(args), std::data(args),
+                            [] { return make_ready_future(); });
```

Comment:
Maybe add a trivial check here that the reactor didn't silently bork.

```c++
const auto int64_t canary_magic = 76868676876;
int64_t canary = 0;
int exit_code = app.run(std::size(args), std::data(args), [&] {
    canary = canary_magic;
    return make_ready_future();
});
BOOST_REQUIRE_EQUAL(canary, canary_magic);
```

Reply:
i tweak the test a little bit so it returns the expected `exit_code` to verify that the app template executes the coroutine for us.